File: frontend/src/api/contracts/agent.ts
import type { CompareResponse } from "./compare";
import type { FaqAskResponse } from "./faq";
import type { IntentParseResponse } from "./intent";


export type AgentToolStatusResponse = {
  name: string;
  enabled: boolean;
  description: string;
};

export type AgentPrecheckResponse = {
  status: string;
  summary: string;
  model: string;
  base_url: string | null;
  api_style: string;
  openai_sdk_available: boolean;
  langgraph_available: boolean;
  data_backend: string;
  agent_log_backend: string;
  catalog_total: number;
  warnings: string[];
  tools: AgentToolStatusResponse[];
};

export type AgentToolCallResponse = {
  tool_name: string;
  status: string;
  summary: string;
  input_payload: Record<string, unknown>;
  output_payload: Record<string, unknown>;
};

export type AgentProvidersResponse = {
  route_provider: string;
  intent_provider: string;
  answer_provider: string;
  retrieval_provider: string;
};

export type AgentConversationTurnResponse = {
  user_message: string;
  agent_answer: string;
  route: "shopping" | "faq" | "compare" | "";
  selected_product_ids: string[];
  recommended_product_ids: string[];
};

export type AgentChatResponse = {
  message: string;
  thread_id: string;
  selected_product_ids: string[];
  conversation_context: AgentConversationTurnResponse[];
  route: "shopping" | "faq" | "compare";
  route_reasoning: string;
  final_answer: string;
  warnings: string[];
  tool_calls: AgentToolCallResponse[];
  parsed_intent: IntentParseResponse | null;
  recommended_product_ids: string[];
  faq_result: FaqAskResponse | null;
  compare_result: CompareResponse | null;
  providers: AgentProvidersResponse;
  provider: string;
  model: string;
  graph_runtime: string;
  run_id: string | null;
  persisted: boolean;
};

export type AgentRunDetailResponse = {
  run_id: string;
  thread_id: string;
  created_at: string;
  message: string;
  selected_product_ids: string[];
  conversation_context: AgentConversationTurnResponse[];
  route: "shopping" | "faq" | "compare";
  route_reasoning: string;
  final_answer: string;
  warnings: string[];
  tool_calls: AgentToolCallResponse[];
  parsed_intent: IntentParseResponse | null;
  recommended_product_ids: string[];
  faq_result: FaqAskResponse | null;
  compare_result: CompareResponse | null;
  providers: AgentProvidersResponse;
  provider: string;
  model: string;
  graph_runtime: string;
  persisted: boolean;
};

export type AgentRunSummaryResponse = {
  run_id: string;
  thread_id: string;
  created_at: string;
  message: string;
  route: "shopping" | "faq" | "compare";
  final_answer_preview: string;
  warning_count: number;
  tool_call_count: number;
  selected_product_ids: string[];
  recommended_product_ids: string[];
  provider: string;
  model: string;
};

export type AgentRunListResponse = {
  backend: string;
  items: AgentRunSummaryResponse[];
};

File: backend/app/schemas/agent.py
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.compare import CompareResponse
from app.schemas.faq import FaqAskResponse
from app.schemas.intent import IntentParseResponse


class AgentToolStatus(BaseModel):
    """Precheck-time tool status."""

    name: str
    enabled: bool
    description: str


class AgentPrecheckResponse(BaseModel):
    """Agent precheck result."""

    status: str
    summary: str
    model: str
    base_url: str | None
    api_style: str
    openai_sdk_available: bool
    langgraph_available: bool
    data_backend: str
    agent_log_backend: str
    catalog_total: int
    warnings: list[str]
    tools: list[AgentToolStatus]


class AgentToolCall(BaseModel):
    """Structured tool-call trace for one agent run."""

    tool_name: str
    status: str
    summary: str
    input_payload: dict[str, Any] = Field(default_factory=dict)
    output_payload: dict[str, Any] = Field(default_factory=dict)


class AgentProviders(BaseModel):
    """Provider details used across agent stages."""

    route_provider: str = ""
    intent_provider: str = ""
    answer_provider: str = ""
    retrieval_provider: str = ""


class AgentConversationTurn(BaseModel):
    """One prior turn passed into the current agent run as conversation memory."""

    user_message: str = Field(..., min_length=1)
    agent_answer: str = Field(..., min_length=1)
    route: str = Field(default="")
    selected_product_ids: list[str] = Field(default_factory=list)
    recommended_product_ids: list[str] = Field(default_factory=list)


class AgentChatRequest(BaseModel):
    """Agent chat request."""

    message: str = Field(..., min_length=1, description="Current user message")
    selected_product_ids: list[str] = Field(
        default_factory=list,
        description="Currently selected products from the frontend workbench",
    )
    conversation_context: list[AgentConversationTurn] = Field(
        default_factory=list,
        description="Recent prior turns kept by the frontend session",
    )


class AgentChatResponse(BaseModel):
    """Agent chat result."""

    message: str
    selected_product_ids: list[str] = Field(default_factory=list)
    conversation_context: list[AgentConversationTurn] = Field(default_factory=list)
    route: str
    route_reasoning: str
    final_answer: str
    warnings: list[str]
    tool_calls: list[AgentToolCall]
    parsed_intent: IntentParseResponse | None = None
    recommended_product_ids: list[str] = Field(default_factory=list)
    faq_result: FaqAskResponse | None = None
    compare_result: CompareResponse | None = None
    providers: AgentProviders = Field(default_factory=AgentProviders)
    provider: str
    model: str
    graph_runtime: str
    run_id: str | None = None
    persisted: bool = False


class AgentRunDetailResponse(BaseModel):
    """Persisted agent run detail."""

    run_id: str
    created_at: str
    message: str
    selected_product_ids: list[str] = Field(default_factory=list)
    conversation_context: list[AgentConversationTurn] = Field(default_factory=list)
    route: str
    route_reasoning: str
    final_answer: str
    warnings: list[str] = Field(default_factory=list)
    tool_calls: list[AgentToolCall] = Field(default_factory=list)
    parsed_intent: IntentParseResponse | None = None
    recommended_product_ids: list[str] = Field(default_factory=list)
    faq_result: FaqAskResponse | None = None
    compare_result: CompareResponse | None = None
    providers: AgentProviders = Field(default_factory=AgentProviders)
    provider: str
    model: str
    graph_runtime: str
    persisted: bool = True


class AgentRunSummary(BaseModel):
    """Persisted agent run summary."""

    run_id: str
    created_at: str
    message: str
    route: str
    final_answer_preview: str
    warning_count: int
    tool_call_count: int
    selected_product_ids: list[str] = Field(default_factory=list)
    recommended_product_ids: list[str] = Field(default_factory=list)
    provider: str
    model: str


class AgentRunListResponse(BaseModel):
    """Recent persisted agent runs."""

    backend: str
    items: list[AgentRunSummary] = Field(default_factory=list)
